capsules/src/mac.rs

```diff
@@ -0,0 +1,566 @@
+    pub fn remaining_data_capacity(&self) -> usize {
+        self.buf.len() - radio::PSDU_OFFSET - radio::MFR_SIZE - self.info.length()
+    }
+
+    /// Appends payload bytes into the frame if possible
+    pub fn append_payload(&mut self, payload: &[u8]) -> ReturnCode {
+        if payload.len() > self.remaining_data_capacity() {
+            return ReturnCode::ENOMEM;
+        }
+        let begin = radio::PSDU_OFFSET + self.info.length();
+        self.buf[begin..begin + payload.len()].copy_from_slice(payload);
+        self.info.data_len += payload.len();
+
+        ReturnCode::SUCCESS
+    }
+}
+
+impl FrameInfo {
+    /// Current size of the frame, not including the MAC footer that is added by
+    /// the hardware when it generates the CRC
+    pub fn length(&self) -> usize {
+        self.data_offset + self.data_len
+    }
+}
+
+/// The contract satisfied by an implementation of an IEEE 802.15.4 MAC device.
+/// Any IEEE 802.15.4 MAC device should expose the following high-level
+/// functionality:
+///
+/// - Configuration of addresses and transmit power
+/// - Preparing frames (data frame, command frames, beacon frames)
+/// - Transmitting and receiving frames
+///
+/// Outlining this in a trait allows other implementations of MAC devices that
+/// divide the responsibilities of software and hardware differently. For
+/// example, a radio chip might be able to completely inline the frame security
+/// procedure in hardware, as opposed to requiring a software implementation.
+pub trait Mac {
+    /// The short 16-bit address of the MAC device
+    fn get_address(&self) -> u16;
+    /// The long 64-bit address (EUI-64) of the MAC device
+    fn get_address_long(&self) -> [u8; 8];
+    /// The 16-bit PAN ID of the MAC device
+    fn get_pan(&self) -> u16;
+    /// The 802.15.4 channel ID of the MAC device
+    fn get_channel(&self) -> u8;
+    /// The transmission power of the MAC device, in dBm
+    fn get_tx_power(&self) -> i8;
+
+    /// Set the short 16-bit address of the MAC device
+    fn set_address(&self, addr: u16);
+    /// Set the long 64-bit address (EUI-64) of the MAC device
+    fn set_address_long(&self, addr: [u8; 8]);
+    /// Set the 16-bit PAN ID of the MAC device
+    fn set_pan(&self, id: u16);
+    /// Set the 802.15.4 channel of the MAC device. `channel` should be a valid
+    /// channel `11 <= channel <= 26`, otherwise EINVAL will be returned
+    fn set_channel(&self, chan: u8) -> ReturnCode;
+    /// Set the transmission power of the MAC device, in dBm. `power` should
+    /// satisfy `-17 <= power <= 4`, otherwise EINVAL will be returned
+    fn set_tx_power(&self, power: i8) -> ReturnCode;
+
+    /// This method must be called after one or more calls to `set_*`. If
+    /// `set_*` is called without calling `config_commit`, there is no guarantee
+    /// that the underlying hardware configuration (addresses, pan ID) is in
+    /// line with this MAC device implementation.
```

Comment:
Got it, so a legal instance of the `Mac` trait could implement `config_commit` as a noop if all the `set_*` methods committed each change.